Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp

```diff
@@ -911,139 +965,250 @@ void IntegratePeaksMD2::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
+        // pos is set equal to mean is useCentroid after each iteration
+        const auto displ = center - prev_pos;
+        auto mdsq = displ.scalar_prod(invCov * displ);
+        if (mdsq > mdsq_max) {
+          // exclude points outside 3 stdevs
+          useEvent = false;
+          nmasked += 1;
+        }
+      }
+
+      // if prev cov_mat chec
+      if (useEvent) {
+        const auto signal = event.second;
+        w_sum += signal;
+
+        if (useCentroid) {
           // update mean
-          for (size_t d = 0; d < mean.size(); ++d) {
-            mean[d] += (signal / w_sum) * (center[d] - mean[d]);
-          }
+          mean += (center - mean) * (signal / w_sum);
+        }
 
-          if (qAxisIsFixed) {
-            // get variance along Q
-            var_Qhat += signal * pow((center[0] - mean[0]), 2);
-          }
-          for (size_t row = 0; row < cov_mat.numRows(); ++row) {
-            for (size_t col = 0; col < cov_mat.numRows(); ++col) {
-              // symmeteric matrix
-              if (row <= col) {
-                double cov = 0.0;
-                if (!qAxisIsFixed) {
-                  cov = signal * (center[row] - mean[row]) *
-                        (center[col] - mean[col]);
-                } else {
-                  cov = signal * (center[row + 1] - mean[row + 1]) *
-                        (center[col + 1] - mean[col + 1]);
-                }
-                if (row == col) {
-                  cov_mat[row][col] += cov;
-                } else {
-                  cov_mat[row][col] += cov;
-                  cov_mat[col][row] += cov;
-                }
+        // weight for variance
+        auto wi = signal * (w_sum - signal) / w_sum;
+        size_t istart = 0;
+        if (qAxisIsFixed) {
+          // variance along Q (skipped in next nested loops below)
+          cov_mat[0][0] += wi * pow((center[0] - mean[0]), 2);
+          istart = 1;
+        }
+        for (size_t row = istart; row < cov_mat.numRows(); ++row) {
+          for (size_t col = istart; col < cov_mat.numRows(); ++col) {
+            // symmeteric matrix
+            if (row <= col) {
+              auto cov =
+                  wi * (center[row] - mean[row]) * (center[col] - mean[col]);
+              if (row == col) {
+                cov_mat[row][col] += cov;
+              } else {
+                cov_mat[row][col] += cov;
+                cov_mat[col][row] += cov;
               }
             }
           }
         }
       }
     }
-    box->releaseEvents();
-  } while (MDiter.next());
-  // normalise the covariance matrix
-  cov_mat /= w_sum;                // normalise by sum of weights
-  cov_mat.Diagonalise(Evec, Eval); // 3 x 3 matrices
-  eigenvals = Eval.Diagonal();
+    // normalise the covariance matrix
+    cov_mat /= w_sum; // normalise by sum of weights
+
+    // check if another iteration is required
+    bool anyMasked = (nIter > 1) ? (nmasked > 0) : true;
```

Comment:
Should this be nIter>0?

Reply:
Yes that's correct - I changed nIter to start from 0 but forgot to change this - thanks for spotting it!